### ISSUE TITLE
[Feature Request]: Add an optional id parameter to the ChatMessage object.

### ISSUE THREAD

**Issue by gich2009** [run-llama/llama_index#14881]
### Feature Description

Add an id parameter to this class that could optionally be passed in or set by default:
'''
class ChatMessage(BaseModel):
    """Chat message."""

    role: MessageRole = MessageRole.USER
    content: Optional[Any] = ""
    additional_kwargs: dict = Field(default_factory=dict)

    def __str__(self) -> str:
        return f"{self.role.value}: {self.content}"

    @classmethod
    def from_str(
        cls,
        content: str,
        role: Union[MessageRole, str] = MessageRole.USER,
        **kwargs: Any,
    ) -> "ChatMessage":
        if isinstance(role, str):
            role = MessageRole(role)
        return cls(role=role, content=content, **kwargs)

    def _recursive_serialization(self, value: Any) -> Any:
        if isinstance(value, (V1BaseModel, V2BaseModel)):
            return value.dict()
        if isinstance(value, dict):
            return {
                key: self._recursive_serialization(value)
                for key, value in value.items()
            }
        if isinstance(value, list):
            return [self._recursive_serialization(item) for item in value]
        return value

    def dict(self, **kwargs: Any) -> dict:
        # ensure all additional_kwargs are serializable
        msg = super().dict(**kwargs)

        for key, value in msg.get("additional_kwargs", {}).items():
            value = self._recursive_serialization(value)
            if not isinstance(value, (str, int, float, bool, dict, list, type(None))):
                raise ValueError(
                    f"Failed to serialize additional_kwargs value: {value}"
                )
            msg["additional_kwargs"][key] = value

        return msg
        '''

### Reason

This is to enable a class user to grab a specific message from a memory object and iterate on that message while discarding all other messages after it. Kind of like how Claude, Openai and other chatbots allow iteration. 

### Value of Feature

It would be great for allowing all sorts of prompt iterations.

**Comment by logan-markewich** [comment]
Can you give more details on what you mean by "allowing all sorts of prompt interactions" ?

**Comment by logan-markewich** [comment]
Just trying to understand the use case a bit more

**Comment by gich2009** [comment]
So the use case here is, I want to be able to edit a message which is already in a llama-index memory object I created. But I don't know an efficient way of doing this if the messages in the memory object do not have unique references. If I could pass in an id (which I have stored somewhere in my database), then I could grab it from the memory and discard all messages after it(sort of like branching out). Although I could also just search all the user messages in the memory and check whether they're equal to the one I'm looking for.


This occurred to me when one of our users asked me if they could edit a prompt and I was at odds on how to implement it.

**Comment by gich2009** [comment]
If you have a way around this @logan-markewich, I'd be more than happy to adopt an alternative.

**Comment by logan-markewich** [comment]
So the requirements are two-fold here I think
1. Adding an ID field to messages
2. Adding a get_by_id() or similar method to the memory class

**Comment by gich2009** [comment]
Yes exactly @logan-markewich. Those should be sufficient to solve the problem.

**Comment by colca** [comment]
Posted https://github.com/colca/llama_index/pull/3 for this @logan-markewich
